### DISCUSSION TITLE
Play Media from a Stream or Byte Array?

### DISCUSSION THREAD

**Discussion by elthombre** [SuRGeoNix/Flyleaf#284] [Q&A] [answered]
Hi,

I need a media player that I can embed on a WinForm wherein I can play media (both audio & video) from a stream or byte array. Can this control do that?

Thanks!

**Comment by SuRGeoNix** [comment, marked as answer]
Hi @elthombre,

Flyleaf accepts an System.IO.Stream as input which will pass down to FFmpeg as an AVIOContext. This will be used by FFmpeg to demux the data. In your case you could force a rawvideo for example and from your Stream just pass the raw video data. (which can be done without a custom Stream, this is just an example)

To play both audio/video ideally you should use a proper FFmpeg format that contains both audio/video data. But if you can't there is still a solution by creating a second demuxer and adding the audio stream to the main (video) demuxer. I'm doing this with yt-dlp plugin for example (different demux url for audio). I have to check but I think you can do this only with a plugin and not directly.

Have a look and when you have a more information on what you want to do, I could help you more as well or write you a sample (eg. for plugin).

J

**Reply by elthombre** [reply]
Hi, thanks for the quick response. What we're doing is working with email attachments. The files will come from there and we can't save them to a folder (i.e. provide a file path). Consequently, once the data is loaded into a byte array or stream, we want to play it with a media player.

**Reply by SuRGeoNix** [reply]
So your data will be actually two byte arrays one containing the video and one containing the audio right?

**Comment by elthombre** [comment]
no. they come in the form of media files. think file extension type but pushing a file up into memory as a byte array or streaming it.

And imagine a Play() method which may accept a byte array or a stream.

**Reply by SuRGeoNix** [reply]
OK so let's say is just an .mp4 format file with both audio/video but you don't have all the data in once. So a single Stream of byte data which contains (partially) both audio/video? Or two different files?

**Comment by elthombre** [comment]
Well, an attachment file will not come as two files.

**Reply by SuRGeoNix** [reply]
OK just trying to make sure what are you dealing with. If FFmpeg can play the format without the need to read eg. last parts of the file or random positions then you can just use the Player.Open with the System.IO.Stream and your byte array. Consider disabling find stream info from player's config to ensure it will not require random position readings.

**Reply by elthombre** [reply]
@SuRGeoNix , are these assemblies strong named, by chance?

**Reply by SuRGeoNix** [reply]
No they are not

**Reply by elthombre** [reply]
[sigh] then they will conflict with our project

**Comment by elthombre** [comment]
Hi John,
I have removed the assembly signing for development. It's a project that I inherited. 
I've also been looking at VLC which has proven tedious, in my opinion. It seems to require the copying of files & folders to the output folder, etc.
Flyleaf gives the impression that it is easier to use, no? 
I have an MS Office add-in. Not an executable. I plan to leverage the player on a WinForm and use a byte array for data so that we do not have to save files to a temp folder (security issue).
Does this sound like something Flyleaf would easily accommodate?
Thanks,
-- Todd

    On Wednesday, April 26, 2023 at 09:56:19 AM EDT, John Stamatakis ***@***.***> wrote:  
 
 


No they are not

**Reply by SuRGeoNix** [reply]
Hi @elthombre, yes sounds it can. If you need something specific just open an issue/enhancement request and I will include it.

**Reply by elthombre** [reply]
@SuRGeoNix - what I need is the following:

-- There are three assemblies offered through Nuget, which should I install for a WinForms solutions?
-- I need a drag-n-drop control, ideally, that I can place on a WinForm and size it.
-- For source / pseudo code working sample of what I need to do:

using Flyleaf_Lib;

        private FlyleafLib libFly;
        private MediaPlayer mp;

        private readonly byte[] FileBytes;

        public Form1()
        {
            InitializeComponent();            

            libFly = new FlyleafLib();
            mp = new MediaPlayer(libFly);
            videoView1.MediaPlayer = mp;
        }

        private void Form1_Load(object? sender, EventArgs e)
        {
            var stream = new MemoryStream(this.FileBytes);

            var media = new Media(FlyleafLib, new StreamMediaInput(stream));
            mp.Play(media);
            media.Dispose();
        }

**Reply by SuRGeoNix** [reply]
Check the Samples and Wiki (You just need FlyleafLib and not WPF/WinUI packages of course)

**Comment by elthombre** [comment]
Getting this exception:

System.NullReferenceException
  HResult=0x80004003
  Message=Object reference not set to an instance of an object.
  Source=FlyleafLib
  
  On this code:
  
              Config = new Config();
            Config.Video.BackgroundColor = System.Windows.Media.Colors.DarkGray;

and I've also tried:

            FlyleafLib.Config config = new Config();
            config.Video.BackgroundColor = System.Windows.Media.Colors.DarkGray;

**Reply by SuRGeoNix** [reply]
Make sure you start the Engine before you use the FlyleafLib config

**Reply by elthombre** [reply]
I went with the sample at the bottom of this page:

https://github.com/SuRGeoNix/Flyleaf/wiki#quick-player-setup-winui-3

**Reply by SuRGeoNix** [reply]
WinUI3 is not WinForms

**Reply by elthombre** [reply]
Then you don't really know what's in your own website. If you scroll to the bottom of the page as stated, it says the following:

"Quick Player Setup (WinForms)"

**Reply by SuRGeoNix** [reply]
You give a link with the WinUI3 instead of this https://github.com/SuRGeoNix/Flyleaf/wiki#quick-player-setup-winforms
Stop being a smartass

**Reply by elthombre** [reply]
@SuRGeoNix - sorry, wasn't trying to be smart. I was surprised that you said it wasn't 'WinForms' was all.

Thanks for the other links. Will try.

**Comment by elthombre** [comment]
I suppose I should also ask about the stream. I have the following:

            //byte array
            var bytes = this.FileBytes;
            var stream = new MemoryStream(bytes);

            this.Player.Open(stream);
            this.Player.Play();
            this.Player.Dispose();

**Reply by SuRGeoNix** [reply]
I guess you should dispose with when the player is stopped/ended check the property changed for Player.Status

**Comment by elthombre** [comment]
working from the samples here:

https://github.com/SuRGeoNix/Flyleaf/wiki#quick-player-setup-winui-3

https://github.com/SuRGeoNix/Flyleaf/wiki/Player-(Open)

**Reply by SuRGeoNix** [reply]
I though you use WinForms? But yes more or less check the Wiki and mainly the solution with Samples for code that will help you start easier.

**Reply by elthombre** [reply]
Yes, it is a WinForms solutions. All I need is a control on a WinForm and then the ability to pass it a byte array. That's it.

I wish that you had a sample like that.

**Reply by SuRGeoNix** [reply]
FlyleafPlayer (Custom) (WinForms) does exactly that it uses a filestream to play the data for the 2nd player

```csharp
Stream customInput = new FileStream(SampleVideo, FileMode.Open);
Player2.OpenAsync(customInput);
```

Did  you even check?

**Reply by elthombre** [reply]
I saw the following:

https://github.com/SuRGeoNix/Flyleaf/blob/master/Samples/FlyleafAudioPlayer%20(Custom)%20(WinForms)/FrmMain.cs

which seemed excessive and went to the "Quick Player Setup" which you claim is *not* WinForms even though the bottom of the page says that it is.
